### PR TITLE
Feature/cdf/american donut

### PR DESCRIPTION
**Pull request by lisakoolen**
Thank you for your contribution to the **Tomni** package.
Before submitting/accepting this Pull Request, please make sure:

## Contributor of code

- [ ] Version number is updated
- [ ] `HISTORY.rst` is updated
- [ ] Function is tested
- [ ] `(docstring)` is added or updated
- [ ] All needed package are in [requirements.txt](requirements.txt)

**For new function**

- [ ] Functions are named based on [Python naming conventions](https://visualgit.readthedocs.io/en/latest/pages/naming_convention.html)
- [ ] Function is added to apidocx.rst
- [ ] New feature is added to `README.md` subsection `Features`
- [ ] New function is imported in all `__init__.py` that are needed

**For bugfix**
- [ ] Unit test is added that covers the bug (if possible)
- [ ] You cursed at the person that introduced the bug

## Reviewer

- [ ] Version number is updated
- [ ] `HISTORY.rst` is updated
- [ ] Function is tested
- [ ] `(docstring)` is added or updated
- [ ] All needed package are in [requirements.txt](requirements.txt)

**For new function**

- [ ] Functions are named based on [Python naming conventions](https://visualgit.readthedocs.io/en/latest/pages/naming_convention.html)
- [ ] Function is added to apidocx.rst
- [ ] New feature is added to `README.md` subsection `Features`
- [ ] New function is imported in all `__init__.py` that are needed

**For bugfix**
- [ ] Unit test is added that covers the bug (if possible)

### CODE REVIEW

File: tomni/annotation_manager/utils/main.py [hunk 1]
```diff
@@ -9,3 +9,14 @@ def parse_points_to_contour(points: List[Point]) -> np.ndarray:
     contour_points = [[[point.x, point.y]] for point in points]
     contour = np.array(contour_points, dtype=np.int32)
     return contour
+
+
+def parse_points_to_inner_contour(
+    list_points: Union[List[List[Point]], List]
+) -> np.ndarray:
```
Comment (Bram-van-der-Velden) [inline tomni/annotation_manager/utils/main.py:16]:
```suggestion
) -> List:
```

File: tomni/annotation_manager/annotations/polygon/main.py [hunk 2]
```diff
@@ -19,60 +18,99 @@
...
+
         """
         super().__init__(id, label, children, parents, accuracy)
-        self._points: List[Point] = simplify_line(points)
+        self._points = points
+        self._inner_points = inner_points
+        if len(self._points) < 5:
+            raise ValueError("Polygon must have atleast 5 points.")
         self._contour: np.ndarray = parse_points_to_contour(points)
+        self._inner_contour: np.ndarray = parse_points_to_inner_contour(inner_points)
```
Comment (Bram-van-der-Velden) [inline tomni/annotation_manager/annotations/polygon/main.py:54]: inner contourS

File: tomni/annotation_manager/annotations/polygon/main.py [hunk 3]
```diff
@@ -222,16 +310,21 @@ def to_binary_mask(self, shape: Tuple[int, int]) -> np.ndarray:
         return mask
 
     def _calculate_area(self) -> None:
-        self._area = cv2.contourArea(self._contour)
+        # outer area for calculating circularity f.e.
+        self._outer_area = cv2.contourArea(self._contour)
+        self._area = self._outer_area
```
Comment (Bram-van-der-Velden) [inline tomni/annotation_manager/annotations/polygon/main.py:315]: just do self._area in 1 go and substract the inner areas instead of renaming a variable

Comment (Bram-van-der-Velden) [reply tomni/annotation_manager/annotations/polygon/main.py:315]: or have a temp area and then at the end self._area = temp_variable

File: tomni/annotation_manager/annotations/polygon/main.py [hunk 4]
```diff
@@ -19,60 +18,99 @@
...
+        self._inner_points = inner_points
+        if len(self._points) < 5:
+            raise ValueError("Polygon must have atleast 5 points.")
         self._contour: np.ndarray = parse_points_to_contour(points)
+        self._inner_contour: np.ndarray = parse_points_to_inner_contour(inner_points)
+        self._metric_unit = metric_unit
+        self._pixel_density = pixel_density
+
+        self._area: Union[float, None] = None
+        self._area_outer: Union[float, None] = None
```
Comment (Bram-van-der-Velden) [inline tomni/annotation_manager/annotations/polygon/main.py:59]: says area_outer here but everywhere else its _outer_area

**Review by Bram-van-der-Velden — CHANGES_REQUESTED** [review]
comments

File: tomni/annotation_manager/test_annotation_manager.py [hunk 5]
```diff
@@ -674,107 +680,542 @@ def test_binary_mask_multiple_objects_not_connected_to_labeled_mask_connectivity
...
+
+        self.assertRaises(
+            AssertionError,
+            AnnotationManager.from_labeled_mask,
+            data,
+            labels=labels,
+            include_inner_contours=True,
+        )
+
+    def test_labeled_mask_classes_mismatch_2(self):
```
Comment (jellevankerk) [inline tomni/annotation_manager/test_annotation_manager.py:1198]: this should not fail. if model expects to predict 2 labels but on 1 image it only sees one label it should not fial

**Review by jellevankerk — APPROVED** [review]

**Pull request merged**